apps/evm_family/evm_sign_txn.c

```diff
@@ -0,0 +1,497 @@
+
+/**
+ * @brief Signs the user verified transaction and returns the signature.
+ * @details The function internally calls wallet reconstruction sub-flow to get
+ * access to the seed. The function will ensure clearing of all the keys
+ * and other sensitive data (such as seed, HDNode, etc.) before exiting and
+ * should ensure no leakage of secret data.
+ *
+ * @param sig Reference to the struct having storage for signature components
+ *
+ * @return bool Indicating if signature for the provided transaction was
+ * successfully generated
+ * @retval true If all the signature is generated without any error
+ * @retval false If any of the singature failed to generate
+ */
+static bool sign_transaction(evm_sign_txn_signature_response_t *sig);
+
+/**
+ * @brief Sends the generated signature to the host
+ *
+ * @param query Reference to an instance of evm_query_t to store transient
+ * request from the host
+ * @param sig Reference to the struct having the signature components
+ *
+ * @return bool Indicating if the signature is sent to the host
+ * @retval true If the signature was sent to host successfully
+ * @retval false If the host responded with unknown/wrong query
+ */
+static bool send_signature(evm_query_t *query,
+                           evm_sign_txn_signature_response_t *sig);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static evm_txn_context_t *txn_context = NULL;
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+static bool check_which_request(const evm_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->sign_txn.which_request) {
+    evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request_data(const evm_sign_txn_request_t *request) {
+  bool status = true;
+
+  if (!evm_derivation_path_guard(request->initiate.derivation_path,
+                                 request->initiate.derivation_path_count)) {
+    evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+
+  /* hard limit on the transaction size that can be fetched and kept in RAM.
+   * this limit can be removed once the RLP decoder can operate on running
+   * transaction buffer stream and get user confirmations on the go
+   */
+  // TODO: pending; currently ineffective
+  if (EVM_TRANSACTION_SIZE_CAP < txn_context->transaction_size) {
```

Comment:
How is `txn_context->transaction_size` populated?

Reply:
it's a todo. will be replaced by a member of `txn_context->init_info`

Reply:
There are two ways to get the total size of the txn/msg data.

1. Current method: 
    From first chunk (total size = chunk size + remaining size)
2. Change in proto required:  
    Get the size from the initiate request from Sdk.

Reply:
> it's a todo. will be replaced by a member of `txn_context->init_info`

Are you suggesting the 2nd method here?

Reply:
yes

Reply:
added all the necessary changes for this